### PR TITLE
Improve handling of read-only files

### LINKED ISSUE #3407: Losing Passwords Due To Read Only Database
## Expected Behavior
When using Keepassxc on my new OS installation with a copy of my database, I continued savings passwords using the "automatically save changes" setting.

## Current Behavior
Due to not thinking to manually change the owner and group on the database, it was opening as read-only. While I was adding new entries for services or machines, they were not being saved. I didn't realize the problem until I'd entered 5 or 6 new entries, and had to go back and recover all of the passwords.

## Possible Solution
Although the database was accurately labeled as "Read Only", it isn't a very eye-catching notification. I like "save automatically" because I don't have to think about checking to make sure my data is saved. Maybe if a database is opened as Read Only *and* the "Save Automatically" setting is set to true, provide some sort of splashy warning to make sure users understand their information is not being saved before they exit?

## Steps to Reproduce
1. Set "save changes automatically" to true.
2. Open a database as read only.
3. Make changes to the database.
4. Close the database.
5. When you reopen the database, changes are gone.

## Debug Info
KeePassXC - 2.4.3

Libraries:
- LIBS

Operating system: Ubuntu
CPU architecture: 86_64

Enabled extensions:
- None

### PR DESCRIPTION
**Pull request by droidmonkey**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change which fixes an issue)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
* Fix #3407
* Read-only files now disable auto-save and show as modified correctly. This allows the GUI to prompt to "save-as" instead of silently discarding changes when the read-only database is locked or closed.

_Note: The assert changes in Database.cpp are never hit due to correct handling of files in DatabaseWidget.cpp. These are in place in case of future invalid handling of read-only files._

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tested using read-only directory on Windows 10.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation, and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

File: src/core/Database.cpp [hunk 1]
```diff
@@ -202,8 +202,12 @@ bool Database::save(QString* error, bool atomic, bool backup)
  */
 bool Database::save(const QString& filePath, QString* error, bool atomic, bool backup)
 {
-    Q_ASSERT(!m_data.isReadOnly);
-    if (m_data.isReadOnly) {
+    // Disallow saving to the same file if read-only
+    if (m_data.isReadOnly && filePath == m_data.filePath) {
+        Q_ASSERT(false);
```
Comment (phoerious) [inline src/core/Database.cpp]: I don't like this. You should at least use Q_ASSERT_X to give a future bug hunter some clue as to what went wrong.

File: src/gui/DatabaseWidget.cpp [hunk 2]
```diff
@@ -984,7 +984,10 @@ void DatabaseWidget::unlockDatabase(bool accepted)
     }
     replaceDatabase(db);
     if (db->isReadOnly()) {
-        showMessage(tr("File opened in read only mode."), MessageWidget::Warning, false, -1);
+        showMessage(tr("This database is opened in read only mode. Auto-save has been disabled."),
```
Comment (phoerious) [inline src/gui/DatabaseWidget.cpp]: "read-only" and "Autosave".

Perhaps also "is disabled" instead of "has been disabled"?

**Comment by droidmonkey** [comment]
Requested changes made

**Comment by zv3rob0y** [comment]
I have lost a VERY important password with default keepassXC 2.4.3 on linux mint :((((
Losing Passwords Due To Read Only Database. Is there any way to recover the password?

**Pull request merged**